aten/src/ATen/core/LegacyTypeDispatch.cpp

```diff
@@ -2,6 +2,51 @@
 
 namespace at {
 
+/// NOTE [ Treating Variables as non-Variables in type dispatch ]
+///
+/// Previously, in VariableType_*.cpp (generated by gen_variable_type.py), when
+/// a function is using the 'use_derived' strategy, we call its implementation
+/// on the base non-Variable type (`baseType`), passing unwrapped tensors to the
+/// call so that any `.type()` calls in the implementation can treat the passed
+/// tensors as non-Variables and won't dispatch back to functions in VariableType.
+///
+/// However, after the Variable/Tensor merge, there is no concept of unwrapping
+/// a tensor anymore, and directly passing variables to the base type calls will
+/// cause the `.type()` dispatch in the implementation to treat the tensor as a
+/// variable, and any function dispatch based on `.type()` will dispatch back to
+/// VariableType, which is not what we want.
```

Comment:
I was explaining this scheme to @zdevito yesterday, and I realized this is not quite accurate. Technically, you can pass the variable to the base type call "unwrapped" by making another shallow copy of the tensor with autograd metadata nulled out. I failed to convince Zach that there wasn't a way for not crazy code (e.g., code that isn't resizing variables) to observe the difference. Do you remember if there was another reason why this didn't work?

Reply:
`I failed to convince Zach that there wasn't a way for not crazy code (e.g., code that isn't resizing variables) to observe the difference` -- sorry having a bit of difficulty parsing all the negatives, can you restate this?

Also, at a minimum wouldn't this require moving all of our autograd related code to ATen from torch (because now you can use type in order to differentiate (hehe), but you can't do that if only the autograd metadata is different.  Not that that is necessarily a bad idea, but I certainly wouldn't want to do it in this patch.

Reply:
Put differently, code that resizes variables can observe the difference if you shallow copy, but we think that code is crazy. Is there normal code that can see the difference?

Anyway, I still think this approach is right, I just wish I remembered all the reasons why :)

Reply:
Functions such as `cholesky_out` / `inverse_out` / `gesv_out` in `BatchLinearAlgebra.cpp` do perform in-place resize on the returned tensor (we could do another resize/copy to get the returned data into the original tensor, but it would make those code paths slower). Other than those functions I think the unwrap-by-shallow-copy approach can work, although the shallow copy step might put some overhead in all of VariableType functions that have base type dispatch.

Reply:
Yes, I think the performance consideration is a compelling enough reason by itself. Just wondering if there were other reasons.